// @react95/core/esm/List/ListItem.js
var _excluded = ["icon", "children"];
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
import React, { forwardRef } from 'react';
import styled, { css } from '@xstyled/styled-components';
import rightcaret from '../GlobalStyle/imgs/rightcaret.svg';
var Item = styled.li.withConfig({
  displayName: "ListItem__Item",
  componentId: "sc-1k1qq02-0"
})(["\n  position: relative;\n\n  display: flex;\n  align-items: center;\n  margin: 0;\n  padding: 2 6;\n\n  color: materialText;\n\n  svg {\n    margin-right: 10;\n  }\n\n  &:hover {\n    background-color: headerBackground;\n    color: materialTextInvert;\n  }\n\n  ul {\n    display: none;\n    position: absolute;\n    top: -2px;\n    left: 97%;\n    color: materialText;\n    z-index: taskbar;\n\n    svg {\n      width: 16px;\n      height: 16px;\n      margin-right: 6;\n    }\n  }\n\n  ", ";\n\n  ", ";\n"], function (_ref) {
  var icon = _ref.icon;
  return !icon && 'padding-left: 26px;';
}, function (_ref2) {
  var hasList = _ref2.hasList;
  return hasList && css(["\n      &:after {\n        position: absolute;\n        width: 5px;\n        height: 8px;\n        right: 8px;\n\n        content: '';\n        background-color: materialText;\n        mask-image: url('", "');\n        mask-position: center center;\n        mask-size: 5px 8px;\n        mask-repeat: no-repeat;\n\n        svg {\n          fill: materialTextInvert;\n        }\n      }\n\n      &:hover {\n        &:after {\n          background-color: materialTextInvert;\n        }\n\n        ul {\n          display: block;\n        }\n      }\n    "], rightcaret);
});
var ListItem = /*#__PURE__*/forwardRef(function (_ref3, ref) {
  var icon = _ref3.icon,
    _ref3$children = _ref3.children,
    children = _ref3$children === void 0 ? [] : _ref3$children,
    rest = _objectWithoutProperties(_ref3, _excluded);
  return /*#__PURE__*/React.createElement(Item, _extends({}, rest, {
    icon: !!icon,
    ref: ref,
    hasList: Boolean(children && React.Children.map(children, function (child) {
      return /*#__PURE__*/React.isValidElement(child);
    }).some(function (child) {
      return child;
    })),
    className: "default"
  }), icon, children);
});
ListItem.displayName = 'List.Item';
export default ListItem;
