// @react95/core/cjs/RadioButton/RadioButton.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireWildcard(require("@xstyled/styled-components"));
var _radioButtonStates = require("./radioButtonStates");
var _excluded = ["children", "disabled"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var Icon = _styledComponents["default"].span.withConfig({
  displayName: "RadioButton__Icon",
  componentId: "sc-1nrg9a4-0"
})(["\n  width: 12px;\n  height: 12px;\n\n  content: '';\n  display: inline-block;\n\n  position: absolute;\n  left: 0;\n  top: 0;\n\n  background-image: url(", ");\n"], _radioButtonStates.radioUnchecked);
var Text = _styledComponents["default"].span.withConfig({
  displayName: "RadioButton__Text",
  componentId: "sc-1nrg9a4-1"
})(["\n  padding: 1;\n  user-select: none;\n\n  position: absolute;\n  top: 0;\n  left: 18px;\n"]);
var Field = _styledComponents["default"].input.attrs({
  type: 'radio'
}).withConfig({
  displayName: "RadioButton__Field",
  componentId: "sc-1nrg9a4-2"
})(["\n  margin: 0;\n  opacity: 0;\n\n  &:focus ~ ", ", &:active + ", " {\n    border-width: 1;\n    border-style: dotted;\n    padding: 0;\n  }\n\n  &:checked + ", " {\n    background-image: url(", ");\n  }\n\n  &:disabled + ", " {\n    background-image: url(", ");\n  }\n\n  &:checked:disabled + ", " {\n    background-image: url(", ");\n  }\n"], Text, Text, Icon, _radioButtonStates.radioChecked, Icon, _radioButtonStates.radioUncheckedDisabled, Icon, _radioButtonStates.radioCheckedDisabled);
var Label = _styledComponents["default"].label.withConfig({
  displayName: "RadioButton__Label",
  componentId: "sc-1nrg9a4-3"
})(["\n  position: relative;\n  margin-bottom: 10;\n  display: block;\n\n  ", "\n"], function (_ref) {
  var disabled = _ref.disabled;
  return disabled && (0, _styledComponents.css)(["\n      color: materialTextDisabled;\n      text-shadow: 0.5px 0.5px ", ";\n    "], (0, _styledComponents.th)('colors.materialTextDisabledShadow'));
});
var RadioButton = /*#__PURE__*/(0, _react.forwardRef)(function (_ref2, ref) {
  var children = _ref2.children,
    disabled = _ref2.disabled,
    props = _objectWithoutProperties(_ref2, _excluded);
  return /*#__PURE__*/_react["default"].createElement(Label, {
    disabled: disabled
  }, /*#__PURE__*/_react["default"].createElement(Field, _extends({
    disabled: disabled
  }, props, {
    ref: ref
  })), /*#__PURE__*/_react["default"].createElement(Icon, null), /*#__PURE__*/_react["default"].createElement(Text, null, children));
});
RadioButton.defaultProps = {
  children: '',
  disabled: false
};
var _default = exports["default"] = RadioButton;
