// @react95/core/cjs/Modal/Modal.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _styledComponents = _interopRequireWildcard(require("@xstyled/styled-components"));
var React = _interopRequireWildcard(require("react"));
var _reactDraggable = _interopRequireDefault(require("react-draggable"));
var _Button = _interopRequireDefault(require("../Button"));
var _TitleBar = _interopRequireDefault(require("../TitleBar"));
var _ModalContext = _interopRequireDefault(require("./ModalContext"));
var _excluded = ["hasWindowButton", "buttons", "buttonsAlignment", "children", "closeModal", "defaultPosition", "positionOffset", "height", "icon", "menu", "title", "width"];
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var ModalWrapper = _styledComponents["default"].div.withConfig({
  displayName: "Modal__ModalWrapper",
  componentId: "sc-4ouk59-0"
})(["\n  display: flex;\n  flex-direction: column;\n\n  position: fixed;\n\n  padding: 2 2 8;\n\n  top: 50px;\n\n  background-color: material;\n\n  box-shadow: inset 1px 1px 0px 1px ", ",\n    inset 0 0 0 1px ", ",\n    1px 1px 0 1px ", ";\n\n  ", "\n  ", "\n"], (0, _styledComponents.th)('colors.borderLightest'), (0, _styledComponents.th)('colors.borderDark'), (0, _styledComponents.th)('colors.borderDarkest'), function (_ref) {
  var width = _ref.width,
    height = _ref.height;
  return "\n    width: ".concat(width ? "".concat(width, "px") : 'auto', ";\n    height: ").concat(height ? "".concat(height, "px") : 'auto', ";\n  ");
}, function (_ref2) {
  var active = _ref2.active;
  return active ? (0, _styledComponents.css)(["\n          z-index: modal;\n        "]) : '';
});
var Content = _styledComponents["default"].div.withConfig({
  displayName: "Modal__Content",
  componentId: "sc-4ouk59-1"
})(["\n  flex-grow: 1;\n  display: flex;\n  flex-direction: column;\n\n  padding: 6;\n"]);
var ButtonWrapper = _styledComponents["default"].div.withConfig({
  displayName: "Modal__ButtonWrapper",
  componentId: "sc-4ouk59-2"
})(["\n  display: flex;\n  flex-direction: row;\n  justify-content: ", ";\n\n  padding: 0 6 6 6;\n\n  & ", " {\n    margin-right: 6;\n    min-width: 70px;\n\n    &:last-child {\n      margin-right: 0;\n    }\n  }\n"], function (_ref3) {
  var _ref3$buttonsAlignmen = _ref3.buttonsAlignment,
    buttonsAlignment = _ref3$buttonsAlignmen === void 0 ? 'center' : _ref3$buttonsAlignmen;
  return buttonsAlignment;
}, _Button["default"]);
var MenuWrapper = _styledComponents["default"].ul.withConfig({
  displayName: "Modal__MenuWrapper",
  componentId: "sc-4ouk59-3"
})(["\n  display: flex;\n  flex-direction: row;\n\n  list-style: none;\n  margin: 0;\n  padding-left: 0;\n  padding-bottom: 3;\n\n  border-bottom-style: solid;\n  border-width: 1;\n  border-bottom-color: borderDark;\n\n  box-shadow: 0 1px 0 0 ", ";\n"], (0, _styledComponents.th)('colors.borderLighter'));
var MenuItem = _styledComponents["default"].li.withConfig({
  displayName: "Modal__MenuItem",
  componentId: "sc-4ouk59-4"
})(["\n  position: relative;\n  padding-left: 6;\n  padding-right: 6;\n\n  user-select: none;\n\n  ul {\n    position: absolute;\n    left: 0;\n    color: ", ";\n  }\n\n  ", ";\n"], (0, _styledComponents.th)('colors.materialText'), function (_ref4) {
  var active = _ref4.active;
  return active && (0, _styledComponents.css)(["\n      background-color: primary;\n      color: materialTextInvert;\n    "]);
});
MenuItem.displayName = 'MenuItem';
var ModalRenderer = function ModalRenderer(_ref5, ref) {
  var _ref5$hasWindowButton = _ref5.hasWindowButton,
    hasButton = _ref5$hasWindowButton === void 0 ? true : _ref5$hasWindowButton,
    buttons = _ref5.buttons,
    buttonsAlignment = _ref5.buttonsAlignment,
    children = _ref5.children,
    closeModal = _ref5.closeModal,
    defaultPosition = _ref5.defaultPosition,
    positionOffset = _ref5.positionOffset,
    height = _ref5.height,
    icon = _ref5.icon,
    menu = _ref5.menu,
    title = _ref5.title,
    width = _ref5.width,
    rest = _objectWithoutProperties(_ref5, _excluded);
  var _React$useContext = React.useContext(_ModalContext["default"]),
    addWindows = _React$useContext.addWindows,
    removeWindow = _React$useContext.removeWindow,
    updateWindow = _React$useContext.updateWindow,
    setActiveWindow = _React$useContext.setActiveWindow,
    activeWindow = _React$useContext.activeWindow;
  var _React$useState = React.useState(null),
    _React$useState2 = _slicedToArray(_React$useState, 2),
    id = _React$useState2[0],
    setId = _React$useState2[1];
  var _React$useState3 = React.useState(''),
    _React$useState4 = _slicedToArray(_React$useState3, 2),
    menuOpened = _React$useState4[0],
    setMenuOpened = _React$useState4[1];
  var _React$useState5 = React.useState(false),
    _React$useState6 = _slicedToArray(_React$useState5, 2),
    isActive = _React$useState6[0],
    setIsActive = _React$useState6[1];
  React.useEffect(function () {
    if (!id) {
      var newId = addWindows({
        icon: icon,
        title: title,
        hasButton: hasButton
      });
      if (newId) {
        setId(newId);
        setActiveWindow(newId);
      }
    } else {
      updateWindow(id, {
        icon: icon,
        title: title,
        hasButton: hasButton
      });
    }
  }, [id, icon, title, hasButton]);
  React.useEffect(function () {
    return function () {
      if (id) {
        removeWindow(id);
      }
    };
  }, [id]);
  React.useEffect(function () {
    return setIsActive(id === activeWindow);
  }, [id, activeWindow]);
  return /*#__PURE__*/React.createElement(_reactDraggable["default"], {
    handle: ".draggable",
    defaultPosition: defaultPosition,
    positionOffset: positionOffset,
    onMouseDown: id ? function () {
      return setActiveWindow(id);
    } : undefined
  }, /*#__PURE__*/React.createElement(ModalWrapper, _extends({
    width: width,
    height: height
  }, rest, {
    active: isActive,
    ref: ref
  }), /*#__PURE__*/React.createElement(_TitleBar["default"], {
    active: isActive,
    icon: icon,
    title: title,
    className: "draggable"
  }, /*#__PURE__*/React.createElement(_TitleBar["default"].OptionsBox, null, /*#__PURE__*/React.createElement(_TitleBar["default"].Option, null, "?"), /*#__PURE__*/React.createElement(_TitleBar["default"].Option, {
    onClick: closeModal
  }, "X"))), menu && menu.length > 0 && /*#__PURE__*/React.createElement(MenuWrapper, null, menu.map(function (_ref6) {
    var name = _ref6.name,
      list = _ref6.list;
    var active = menuOpened === name;
    return /*#__PURE__*/React.createElement(MenuItem, {
      key: name,
      onMouseDown: function onMouseDown() {
        return setMenuOpened(name);
      },
      active: active
    }, name, active && list);
  })), /*#__PURE__*/React.createElement(Content, {
    onClick: function onClick() {
      return setMenuOpened('');
    }
  }, children), buttons && buttons.length > 0 && /*#__PURE__*/React.createElement(ButtonWrapper, {
    buttonsAlignment: buttonsAlignment
  }, buttons.map(function (button) {
    return /*#__PURE__*/React.createElement(_Button["default"], {
      key: button.value,
      onClick: button.onClick,
      value: button.value
    }, button.value);
  }))));
};
var Modal = /*#__PURE__*/React.forwardRef(ModalRenderer);
Modal.displayName = 'Modal';
Modal.defaultProps = {
  icon: undefined,
  title: 'Modal',
  buttonsAlignment: 'flex-end',
  children: null,
  defaultPosition: {
    x: 0,
    y: 0
  },
  buttons: [],
  menu: [],
  width: undefined,
  height: undefined,
  closeModal: function closeModal() {}
};
var _default = exports["default"] = Modal;
