// @react95/core/esm/Avatar/Avatar.js
var _excluded = ["src", "srcSet", "alt", "circle", "children"];
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
import * as React from 'react';
import styled, { css } from '@xstyled/styled-components';
import Frame from '../Frame/Frame';
var Image = styled.img.withConfig({
  displayName: "Avatar__Image",
  componentId: "sc-1crmc5x-0"
})(["\n  max-width: 100%;\n"]);
var StyledAvatar = styled(Frame).withConfig({
  displayName: "Avatar__StyledAvatar",
  componentId: "sc-1crmc5x-1"
})(["\n  display: flex;\n  align-items: center;\n  justify-content: center;\n\n  font-weight: bold;\n  overflow: hidden;\n  margin-right: 1;\n  margin-bottom: 1;\n\n  box-shadow: unset;\n\n  ", "\n\n  ", "\n"], function (_ref) {
  var circle = _ref.circle,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 48 : _ref$size;
  return css(["\n    border-radius: ", ";\n\n    width: ", ";\n    height: ", ";\n  "], circle ? '50%' : 0, size, size);
}, function (_ref2) {
  var circle = _ref2.circle;
  return circle ? css(["\n          border-style: solid;\n          border-width: 1;\n          border-top-color: borderDark;\n          border-right-color: borderLightest;\n          border-bottom-color: borderLightest;\n          border-left-color: borderDark;\n        "]) : css(["\n          box-shadow: in;\n          padding-left: 1;\n        "]);
});
var Avatar = /*#__PURE__*/React.forwardRef(function (_ref3, ref) {
  var src = _ref3.src,
    srcSet = _ref3.srcSet,
    alt = _ref3.alt,
    circle = _ref3.circle,
    children = _ref3.children,
    otherProps = _objectWithoutProperties(_ref3, _excluded);
  return /*#__PURE__*/React.createElement(StyledAvatar, _extends({
    circle: circle,
    ref: ref
  }, otherProps), src || srcSet ? /*#__PURE__*/React.createElement(Image, {
    src: src,
    srcSet: srcSet,
    alt: alt
  }) : children);
});
export default Avatar;
